src/gc/gc.cpp

```diff
@@ -10602,7 +10602,18 @@ gc_heap::init_gc_heap (int  h_number)
     make_background_mark_stack (b_arr);
 #endif //BACKGROUND_GC
 
-    adjust_ephemeral_limits();
+    ephemeral_low = generation_allocation_start(generation_of(max_generation - 1));
+    ephemeral_high = heap_segment_reserved(ephemeral_heap_segment);
+
+#ifndef MULTIPLE_HEAPS
+    // we have just calculated actual boundaries of ephemeral segment
+    // now it's time to call GCToEEInterface::StompWriteBarrier
+    // which we do only here and only once for wks gc
+    // for MULTIPLE_HEAPS mode we have already stomped write barrier during GCHeap::Initialize()
+    // we are so worried about number of calls to GCToEEInterface::StompWriteBarrier
+    // because it likely flushes instruction cache in each invocation
+    stomp_write_barrier_initialize(ephemeral_low, ephemeral_high);
+#endif //!MULTIPLE_HEAPS
```

Comment:
instead of calling it in 2 places you can just get rid of the one in GCHeap::Initialize and change this to:

```c++
    if (heap_number == 0)
    {
        ephemeral_low = reinterpret_cast<uint8_t*>(1);
        ephemeral_high = reinterpret_cast<uint8_t*>(~0);

#ifndef MULTIPLE_HEAPS
        ephemeral_low = generation_allocation_start(generation_of(max_generation - 1));
        ephemeral_high = heap_segment_reserved(ephemeral_heap_segment);
#endif //!MULTIPLE_HEAPS

        stomp_write_barrier_initialize(ephemeral_low, ephemeral_high);
    }
```

so in Server GC this is only called on heap0.